### PR TITLE
[#11571] Selective deadline extension:  E2E Tests

### PR DESCRIPTION
**Pull request by samuelfangjw**
Part of #11571

**Outline of Solution**
- Add e2e tests for selective deadline extension

### CODE REVIEW

File: src/e2e/java/teammates/e2e/cases/InstructorSessionIndividualExtensionPageE2ETest.java [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+        feedbackSession = testData.feedbackSessions.get("firstSession");
+        students = testData.students.values();
+        instructors = testData.instructors.values();
+
+        removeAndRestoreDataBundle(testData);
+    }
+
+    @Test
+    @Override
+    public void testAll() {
```
Comment (moziliar) [inline src/e2e/java/teammates/e2e/cases/InstructorSessionIndividualExtensionPageE2ETest.java:47]: Should we add:
* a test case where we extend some students and update some other students to a new timestamp X? So a mixed operation.
* a test case where users with and without extension are selected and we revoke extension: it will work on those with extension, and revocation won't be allowed if no extended users are selected

**Review by moziliar — COMMENTED** [review]
I don't have much comments on this PR. Good job!

There are some code duplication that are obvious. Any possible way to reduce?
Looks like `GetDeadlineExtensionData` is tested in this PR too. Could have been a separate PR?

File: src/e2e/java/teammates/e2e/cases/InstructorSessionIndividualExtensionPageE2ETest.java (hunk 1 above)
Comment (samuelfangjw) [reply src/e2e/java/teammates/e2e/cases/InstructorSessionIndividualExtensionPageE2ETest.java:47]: Possible, but this is an e2e to testcase so we should choose which testcases to test carefully. From what I understand, e2e testcase is to verify the overall workflow and that components work well together rather than to test every possible branch. Need to choose some representative test cases rather than trying to cover all cases.

Comment (samuelfangjw) [reply src/e2e/java/teammates/e2e/cases/InstructorSessionIndividualExtensionPageE2ETest.java:47]: To add on, from my understanding, the criteria for justifying a new e2e testcase should be whether the way the user interacts with the system is different. Covering the different edge cases might be better left to unit and integration tests.

Relevant discussion on this: https://github.com/TEAMMATES/teammates/pull/11613#discussion_r817712213

File: src/e2e/resources/data/InstructorSessionIndividualExtensionPageE2ETest.json [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+      "institute": "TEAMMATES Test Institute 1",
+      "timeZone": "Asia/Singapore"
+    }
+  },
+  "instructors": {
+    "ISesIe.instructor1": {
+      "googleId": "tm.e2e.ISesIe.instructor1",
+      "courseId": "tm.e2e.ISesIe.CS2104",
+      "name": "Instructor 1",
+      "email": "instructor1.tmms@gmail.tmt",
```
Comment (jayasting98) [inline src/e2e/resources/data/InstructorSessionIndividualExtensionPageE2ETest.json:31]: This does not match up with the one in accounts. Same for the other instructor and same for student Danny in the feedback responses. Not sure if it's worth bothering with though since the test cases still pass.

**Review by jayasting98 — COMMENTED** [review]
Pretty good work; I don't think there's much that needs to be changed. Just wondering something.

File: src/e2e/resources/data/InstructorSessionIndividualExtensionPageE2ETest.json (hunk 2 above)
Comment (samuelfangjw) [reply src/e2e/resources/data/InstructorSessionIndividualExtensionPageE2ETest.json:31]: Thanks for spotting this, yes this was a mistake on my part.

File: src/e2e/java/teammates/e2e/cases/InstructorSessionIndividualExtensionPageE2ETest.java (hunk 1 above)
Comment (moziliar) [reply src/e2e/java/teammates/e2e/cases/InstructorSessionIndividualExtensionPageE2ETest.java:47]: I see. So the existing consensus is that it should cover the common use cases of the new feature. In that case, I believe this PR has done a great job! Not much to add then:)

Comment (wkurniawan07) [reply src/e2e/java/teammates/e2e/cases/InstructorSessionIndividualExtensionPageE2ETest.java:47]: @moziliar I did mention the same thing to you before: https://github.com/TEAMMATES/teammates/pull/11567#issuecomment-1043686192.

In general, unit tests need to be designed with the mindset that E2E tests don't exist.

**Review by moziliar — APPROVED** [review]
LGTM

**Review by jayasting98 — APPROVED** [review]
I think this should be pretty much good to go.
